core/server/worker/src/main/java/alluxio/worker/netty/DataServerReadHandler.java

```diff
@@ -186,104 +216,94 @@ private boolean tooManyPendingPackets() {
   }
 
   /**
-   * @return true if we should restart the packet reader
-   */
-  @GuardedBy("mLock")
-  private boolean shouldRestartPacketReader() {
-    return !mPacketReaderActive && !tooManyPendingPackets() && mPosToQueue < mRequest.end();
-  }
-
-  /**
-   * @return the number of bytes remaining to push to the netty queue. Return 0 if it is cancelled
-   */
-  @GuardedBy("mLock")
-  private long remainingToQueue() {
-    // mRequest is not guarded by mLock and is volatile. It can be reset because of cancel before
-    // the packet reader is done. We need to make a copy here to make sure the packet reader
-    // thread is not killed due to null pointer exception.
-    ReadRequestInternal request = mRequest;
-    return (request == null || mPosToQueue == -1) ? 0 : request.end() - mPosToQueue;
-  }
-
-  /**
-   * @return the number of bytes remaining to flush. Return 0 if it is cancelled
+   * Returns true if the block read request is valid.
+   *
+   * @param request the block read request
+   * @return the error message (empty string indicates success)
    */
-  @GuardedBy("mLock")
-  private long remainingToWrite() {
-    // mRequest is not guarded by mLock and is volatile. It can be reset because of cancel before
-    // the packet reader is done. We need to make a copy here to make sure the packet reader
-    // thread is not killed due to null pointer exception.
-    ReadRequestInternal request = mRequest;
-    return (request == null || mPosToWrite == -1) ? 0 : request.end() - mPosToWrite;
+  private String validateReadRequest(Protocol.ReadRequest request) {
+    if (request.getId() < 0) {
+      return String.format("Invalid blockId (%d) in read request.", request.getId());
+    }
+    if (!request.getCancel() && (request.getOffset() < 0 || request.getLength() <= 0)) {
+      return String.format("Invalid read bounds in read request %s.", request.toString());
+    }
+    return "";
   }
 
   /**
-   * Writes an error read response to the channel and closes the channel after that.
+   * @param channel the channel
+   * @param error the error
    */
-  private void replyError(Channel channel, Protocol.Status.Code code, String message, Throwable e) {
-    channel.writeAndFlush(RPCProtoMessage.createResponse(code, message, e, null))
-        .addListener(ChannelFutureListener.CLOSE);
+  private void setError(Channel channel, Error error) {
+    Preconditions.checkNotNull(error);
+    mLock.lock();
+    try {
+      if (mError != null) {
```

Comment:
why do the other `setX` methods check `mError != null || mCancel || mEof` and this one does not?